### PR TITLE
Go interfaces for GEP-91

### PR DESCRIPTION
**Pull request by frankbu**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:
This PR adds the interfaces defined in GEP-91 to hopefully help speed up review and approval of #2273.

**Which issue(s) this PR fixes**:

Depends on #2273 

**Does this PR introduce a user-facing change?**:

```release-note
Go interfaces for GEP-91.
```

### CODE REVIEW

File: apis/v1/gateway_types.go [hunk 1]
```diff
@@ -440,6 +449,44 @@ type GatewayTLSConfig struct {
...
+	// +kubebuilder:validation:MinItems=1
+	CACertificateRefs []SecretObjectReference `json:"caCertificateRefs"`
+
+	// SubjectAltNames contains one or more alternate names to verify
+	// the subject identity in the certificate presented by the client.
+	//
+	// Support: Core
+	//
+	// +optional
+	SubjectAltNames []string `json:"subjectAltNames,omitempty"`
```
Comment (howardjohn) [inline apis/v1/gateway_types.go:487]: nit: min=1,max=??

**Review by howardjohn — COMMENTED** [review]
LGTM

File: apis/v1/gateway_types.go (hunk 1 above)
Comment (frankbu) [reply apis/v1/gateway_types.go:488]: Good question about max? I think maybe we should leave it unspecified (it's implementation defined in X.509)?

cc @arkodg, implementation defined or not, I think this needs some update(s) in #2273

Comment (robscott) [reply apis/v1/gateway_types.go:489]: This field has been removed altogether in the latest version of the GEP.

File: apis/v1/gateway_types.go [hunk 2]
```diff
@@ -424,6 +424,15 @@ type GatewayTLSConfig struct {
...
 	CertificateRefs []SecretObjectReference `json:"certificateRefs,omitempty"`
 
+	// ClientValidation can be used to validate the client initiating a TLS connection
+	// to the Gateway.
+	//
+	// Support: Extended
+	//
+	// +optional
+	// <gateway:experimental>
+	ClientValidation *ClientValidationContext `json:"clientValidation,omitempty"`
```
Comment (robscott) [inline apis/v1/gateway_types.go:434]: This is now "FrontendValidation" in the updated GEP

File: apis/v1/gateway_types.go [hunk 3]
```diff
@@ -440,6 +449,46 @@ type GatewayTLSConfig struct {
...
+	//
+	// References to a resource in a different namespace are invalid UNLESS there
+	// is a ReferenceGrant in the target namespace that allows the certificate
+	// to be attached. If a ReferenceGrant does not allow this reference, the
+	// "ResolvedRefs" condition MUST be set to False for this listener with the
+	// "RefNotPermitted" reason.
+	//
+	// +kubebuilder:validation:MaxItems=8
+	// +kubebuilder:validation:MinItems=1
+	CACertificateRefs []SecretObjectReference `json:"caCertificateRefs"`
```
Comment (robscott) [inline apis/v1/gateway_types.go:480]: This has been changed in the latest version of the GEP - we should not be recommending or defaulting to the use of Secrets here.

**Review by robscott — COMMENTED** [review]
Thanks @frankbu! The corresponding GEP PR is approved and merging, do you mind updating this to reflect the updates in that GEP?

**Comment by arkodg** [comment]
chatted with Frank, and I'll pick this up and raise a PR in a bit

**Pull request closed without merging**